compiler/rustc_resolve/src/check_unused.rs

```diff
@@ -38,14 +39,15 @@ use rustc_session::lint::BuiltinLintDiagnostics;
 use rustc_span::symbol::{kw, Ident};
 use rustc_span::{Span, DUMMY_SP};
 
-struct UnusedImport<'a> {
-    use_tree: &'a ast::UseTree,
+#[derive(Clone)]
+struct UnusedImport {
+    use_tree: ast::UseTree,
```

Comment:
> I still don't understand why `UseTree` cloning is necessary here.

Thank you very much.

When I don't change `UnusedImport` and use following code :
```rust
        let unused_imports = visitor.unused_imports;
        let mut check_redundant_imports = FxIndexSet::default();
        for module in self.arenas.local_modules().iter() {
            for (_key, resolution) in self.resolutions(*module).borrow().iter() {
                let resolution = resolution.borrow();

                if let Some(binding) = resolution.binding
                    && let NameBindingKind::Import { import, .. } = binding.kind
                    && let ImportKind::Single { id, .. } = import.kind
                {
                    if let Some(unused_import) = unused_imports.get(&import.root_id)
                        && unused_import.unused.contains(&id)
                    {
                        continue;
                    }

                    check_redundant_imports.insert(import);
                }
            }
        }


```

I got errors :

```bash
error[E0502]: cannot borrow `*self.arenas` as immutable because it is also borrowed as mutable
   --> compiler\rustc_resolve\src\check_unused.rs:489:23
    |
328 |             r: self,
    |                ---- mutable borrow occurs here
...
489 |         for module in self.arenas.local_modules().iter() {
    |                       ^^^^^^^^^^^ immutable borrow occurs here
...
497 |                     if let Some(unused_import) = unused_imports.get(&import.root_id)
    |                                                  -------------- mutable borrow later used here

error[E0499]: cannot borrow `*self` as mutable more than once at a time
   --> compiler\rustc_resolve\src\check_unused.rs:490:39
    |
328 |             r: self,
    |                ---- first mutable borrow occurs here
...
490 |             for (_key, resolution) in self.resolutions(*module).borrow().iter() {
    |                                       ^^^^ second mutable borrow occurs here
...
497 |                     if let Some(unused_import) = unused_imports.get(&import.root_id)
    |                                                  -------------- first borrow later used here

   Compiling rustc_hir_analysis v0.0.0 (D:\source\rust\rust\compiler\rustc_hir_analysis)
   Compiling rustc_lint v0.0.0 (D:\source\rust\rust\compiler\rustc_lint)
Some errors have detailed explanations: E0499, E0502.
For more information about an error, try `rustc --explain E0499`.
error: could not compile `rustc_resolve` (lib) due to 2 previous errors
```

I think because `use_tree` in `UnusedImport` is  borrowed and share the same lifetime  as 'Resolver` in UnusedImportCheckVisitor.